### PR TITLE
663 blueprints match by field and label requirements

### PR DESCRIPTION
**Pull request by idoru**
## Changes proposed by this PR

Introduce `selectorMatchExpressions` and `selectorMatchFields` to `Workload` and `Deliverable`.

This PR introduces support for the `v1alpha1` specifications as per [RFC](https://github.com/vmware-tanzu/cartographer/pull/602)

Updates #663 but does not complete. Work remaining:

* documentation
* possible refactoring (comments inline in code)

## Release Note

Blueprints can now match against fields and use match expressions on labels against `Workload` and `Deliverable`

* Operators can now specify `selectorMatchFields` on `ClusterDelivery` and `ClusterSupplyChain` to match against fields of `Deliverable` and `Workload` candidates
* Operators can now specify `selectorMatchExpressions` on `ClusterDelivery` and `ClusterSupplyChain` to match against labels of `Deliverable` and `Workload` candidates
* At least one element from any of `selector`, `selectorMatchFields` or `selectorMatchExpressions` must be present.
*

## PR Checklist

Note: Please do not remove items. Mark items as done `[x]` or use ~strikethrough~ if you believe they are not relevant

- [x] Linked to a relevant issue. Eg: `Fixes #123` or `Updates #123`
- [x] Filled in the [Release Note](#Release-Note) section above 
- [ ] Modified the docs to match changes

### CODE REVIEW

File: pkg/repository/selector_matcher.go [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+		size := 0
+		ls := &metav1.LabelSelector{
+			MatchLabels:      target.GetMatchLabels(),
+			MatchExpressions: target.GetMatchExpressions(),
+		}
+
+		// -- Labels
+		sel, err := metav1.LabelSelectorAsSelector(ls)
+		if err != nil {
+			// TODO deal with me
```
Comment (cirocosta) [inline pkg/repository/selector_matcher.go]: should we deal with it before the final merge?

File: pkg/repository/selector_matcher.go [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+
+	"github.com/vmware-tanzu/cartographer/pkg/apis/v1alpha1"
+	"github.com/vmware-tanzu/cartographer/pkg/selector"
+)
+
+// * handle errors correctly (panic!)
+
+// Selector interface is an object we can find selectors on, and identify in errors
```
Comment (cirocosta) [inline pkg/repository/selector_matcher.go:30]: intended to keep in the PR or to be addressed later? not sure if a wip

File: pkg/repository/cache.go [hunk 3]
```diff
@@ -109,24 +109,24 @@ func (c *cache) isSubmittedCacheHit(submitted *unstructured.Unstructured, key st
 
 func (c *cache) isPersistedCacheHit(key string, existingObj *unstructured.Unstructured, persistedCached *unstructured.Unstructured) bool {
 	c.logger.Info("considering object", "key", key, "existingName", existingObj.GetName())
-	existingSpec, ok := existingObj.Object["spec"]
+	existingSpec, ok := existingObj.Object["Spec"]
```
Comment (cirocosta) [inline pkg/repository/cache.go]: is this right? I'm not sure how the conversion happens under the hood, but if it's a `map[string]interface{}`, then I believe it's case sensitive (I guess this came in a late refactor?)

File: pkg/repository/selector_matcher.go (hunk 1 above)
Comment (squeedee) [reply pkg/repository/selector_matcher.go]: we said no in the PR notes

File: pkg/repository/selector_matcher.go (hunk 2 above)
Comment (squeedee) [reply pkg/repository/selector_matcher.go:30]: ditto

File: pkg/repository/cache.go (hunk 3 above)
Comment (squeedee) [reply pkg/repository/cache.go]: yeah we're moving too fast, cleaning some strange refactors now.

File: pkg/repository/selector_matcher.go (hunk 2 above)
Comment (squeedee) [reply pkg/repository/selector_matcher.go:30]: Done

File: pkg/repository/selector_matcher.go [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+	}
+
+	return matchingSelectors[highWaterMark], nil
+}
+
+func matchesAllFields(source Selectable, fields []v1alpha1.FieldSelectorRequirement) (bool, error) {
+	for _, requirement := range fields {
+		match, err := selector.Matches(requirement, source)
+		if err != nil {
+			return false, fmt.Errorf("unable to match field requirement with key [%s] operator [%s] values [%v]: %w", requirement.Key, requirement.Operator, requirement.Values, err)
```
Comment (cirocosta) [inline pkg/repository/selector_matcher.go:122]: nit: wdyt of breaking this like lines `73-76`?

**Review by cirocosta — APPROVED** [review]
lgtm!

**Pull request merged**